workflow-templates/test-package.yaml

```diff
@@ -0,0 +1,76 @@
+name: Test
+
+on:
+  # Run CI on pushes to the main branch, and on PRs against main.
+  push:
+    branches: [ $default-branch ]
+  pull_request:
+    branches: [ $default-branch ]
+
+env:
+  PUB_ENVIRONMENT: bot.github
+
+jobs:
+  # Check code formatting and static analysis on a single OS (linux)
+  # against Dart stable and dev.
+  analyze:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        sdk: [dev, stable]
+    steps:
+      - uses: actions/checkout@v2
+      - name: Cache Pub hosted dependencies
```

Comment:
Not sure I get this part; where can Iearn more?

Reply:
This caches the pub downloads:
```yaml
          # Cache this directory (typically everything downloaded from pub.dev)
          path: "~/.pub-cache/hosted"
          # Use this string is as the cache key (the templated values will be filled in when the task is running)
          key: "os:ubuntu-latest;pub-cache-hosted;dart:${{ matrix.sdk }};packages:${{ github.repository }}"
          # Try to find existing caches to initialize this case with, it will use the first match
          restore-keys: |
            # exact match, should have everything we need
            os:${{ matrix.os }};pub-cache-hosted;dart:${{ matrix.sdk }};packages:${{ github.repository }}
            # cache from another run on a build on the same channel, might have some things we need
            os:${{ matrix.os }};pub-cache-hosted;dart:${{ matrix.sdk }}
            # cache from another run on a build on a different channel, might have some things we need
            os:${{ matrix.os }};pub-cache-hosted
            # cache from another run on the same OS, might have some things we need
            os:${{ matrix.os }}
```

Documentation is here: https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows

Reply:
I've removed the caching again. This won't really work as well over time because the caches on GitHub are immutable and these cache keys would persist over a long time and eventually only contain outdated dependencies.

We'd need something like this if we don't want to check-in lock files:
```
pub lock # does not exist, only creates the lock file but doesn't download.
<restore cache with hash of lockfile as cache key>
pub get # would read only from disk because of lockfile and initialized cache or download and initialize cache.
```